src/components/include/cbuf.h

```diff
@@ -405,241 +341,131 @@ __cbuf_alloc_meta_inconsistent(struct cbuf_alloc_desc *d, struct cbuf_meta *m)
-		 * address in memory of the cbuf.  If they do not,
-		 * then we know that the slab is outdated and that the
-		 * cbuf backing it has been taken from this component.
-		 * In that case (shown in the following code), we
-		 * delete the slab descriptor.  Again, see
-		 * cbuf_slab_free to see the surprising fact that we
-		 * do _not_ deallocate the slab descriptor there to
-		 * reinforce that point.
-		 */
-		__cbuf_desc_free(d);
-		if (likely(!already_used)) {
-			cm->nfo.c.flags &= ~CBUFM_TOUCHED;
-			cm->nfo.c.refcnt--;
-		}
+	inconsistent = CBUF_INCONSISENT(cm);
+	if (unlikely(inconsistent)) {
+		CBUFM_SET_NEXT_NULL(cm);
 		goto again;
 	}
-
-	if (tmem) {
+	unsigned int p, n;
+	p = cm->nfo;
+	n = p+CBUFM_INC_UNIT;
+	if (unlikely(!cos_cas((unsigned long *)&cm->nfo, p, n))) {
+		CBUFM_SET_NEXT_NULL(cm);
+		goto again;
+	}
+	CBUFM_SET_NEXT_NULL(cm);
+	cm->owner_nfo.c.nsent = cm->owner_nfo.c.nrecvd = 0;
+	/*if (unlikely(tmem)) {
 		cm->owner_nfo.thdid = cos_get_thd_id();
 		cos_comp_info.cos_tmem_available[COMP_INFO_TMEM_CBUF]--;
 		assert(cm->nfo.c.flags & CBUFM_TMEM);
-	} else {
-		cm->owner_nfo.c.nsent = cm->owner_nfo.c.nrecvd = 0;		
-	}
-	ret = (void*)(cm->nfo.c.ptr << PAGE_ORDER);
+	} */     //???
 done:
-	*cb = cbuf_cons(cbid, len, tmem);
-	CBUF_RELEASE();
+	if (unlikely(tmem)) CBUF_SET_TMEM(cm);
+	ret = (void *)(CBUFM_GET_PTR(cm) << PAGE_ORDER);
+	cbid = cm->cbid.cbid;
+	*cb = cbuf_cons(cbid);
 	return ret;
 }
 
-static inline void *
-cbuf_alloc(unsigned int sz, cbuf_t *cb) { return __cbuf_alloc(sz, cb, 1); }
-static inline void *
-cbufp_alloc(unsigned int sz, cbufp_t *cb)  { return __cbuf_alloc(sz, (cbuf_t*)cb, 0); }
-
 /* 
  * precondition:  cbuf lock must be taken.
  * postcondition: cbuf lock has been released.
  */
 static inline void
-__cbuf_done(int cbid, int tmem)
+cbuf_free(cbuf_t cb)
 {
-	struct cbuf_meta *cm;
-	int owner, relinq = 0;
+	u32_t id;
+	cbuf_unpack(cb, &id);
+	struct cbuf_meta *cm, *fl, head, new_head;
+	int owner, relinq = 0, tmem, inconsistent;
+	unsigned long long *target, *old, *update;
 
-	CBUF_TAKE();
-	cm = cbuf_vect_lookup_addr(cbid_to_meta_idx(cbid), tmem);
+	cm = cbuf_vect_lookup_addr(cbid_to_meta_idx(id));
 	/* 
 	 * If this assertion triggers, one possibility is that you did
 	 * not successfully map it in (cbufp2buf or cbufp_alloc).
 	 */
-	assert(cm->nfo.c.refcnt);
-	owner = cm->nfo.c.flags & CBUFM_OWNER;
-
+	assert(CBUFM_GET_REFCNT(cm));
+	owner = CBUF_OWNER(cm);
+	tmem = CBUF_TMEM(cm);
+	inconsistent = CBUF_INCONSISENT(cm);
+	assert(!inconsistent);
 	if (tmem) {
-		assert(owner); /* Shouldn't be calling free... */
-		struct cbuf_alloc_desc *d = __cbuf_alloc_lookup(cm->nfo.c.ptr);
-		struct cbuf_alloc_desc *fl;
-
-		assert(d && !__cbuf_alloc_meta_inconsistent(d, cm));
-
-		fl = d->flhead;
-		assert(fl);
-		ADD_LIST(fl, d, next, prev);
-		cm->owner_nfo.thdid = 0;
+		assert(owner);
+		fl = __cbuf_freelist_get(cm->sz << PAGE_ORDER);
+		CBUFM_SET_NEXT(cm, CBUFM_GET_NEXT(fl));
+		cos_mem_fence();
```

Comment:
You need to document the shared memory protocol for communication between client and server.  I'll need those to understand these fences.